Support a "default" key for object in scales

folz opened the pull request:
The implementation is straightforward, but I wanted to discuss something that I think is a pretty large change to the meaning of `get` and/or theme paths.

Before this PR, theme-ui does not enforce that `get` on a theme path will return a primitive value suitable for inclusion in CSS. It assumes either that the theme author will write the theme such that all paths lead to primitive values (that is, avoid using objects or arrays at the end of a path), or that the end programmer will ensure all paths will point to a primitive value (that is, `'red.500'`, never `'red'`).

This PR changes the above by moving some of the "ensure path points to a primitive" responsibility to theme-ui, by establishing the convention that `path => object ? object.default`. This is a special case, though. Why not `path => object ? object.500`, or `path => object ? object.core`?  Should we support this for object-like values such as Map? What about other ways to store values: `path => array ? array['default']`? `path => symbol ? symbol.description`?

theme-ui also allows for `sx` to contain functions: `sx={{color: (theme: Theme) => theme.colors.blue.500}}`

This is less ergonomic but more precise than the `object.default` convention. It's also more typesafe, thanks to the `Theme` type. What if we allowed the theme to also contain themeGet functions? `path => function ? function(theme)`. This would allow things like

```
colors: {
  default: themeGet('colors.500'),
  500: 'color500',
  ...
}
```

Is that a reasonable extension of the "ensure path points to a primitive" responsibility?

Anyway, we don't need to answer literally every question above. Maybe the single special case of `object => default` is fine here. But I still wanted to bring up the possible implications to have a discussion about.

Closes #936

folz committed:
[RFC] "default" key for object in scales

jxnblk commented:
Thanks for tackling this! I think this is a great addition to the API as-is -- it also sets up a nice convention for handling default variants.

> What if we allowed the theme to also contain themeGet functions? 

I think this *is* something we'd like to solve for. In the v1 issue, I put this as *Idea: Provide some solution for self-referencing scales in other scale definitions.* I don't think we need to solve for that in this PR, but feel free to take a stab at proposing what that could look like, if you have time

As far as the *ensure path points to a primitive* point, that's something I hadn't really considered, but might be good to at least warn users if an object or invalid value is being passed on as CSS.

jxnblk commented on packages/css/test/index.ts:
```diff
@@ -6,6 +6,17 @@ const theme: Theme = {
     text: 'black',
+    colorScale: {
+      default: 'color500',
+      100: 'color100',
```
Nit: I'd prefer to see valid CSS color values here -- I think it helps with reading the tests and making sure it's doing what you'd expect to happen

jxnblk commented on packages/css/test/index.ts:
```diff
@@ -206,13 +217,44 @@ test('works with functional arguments', () => {
+    color: 'colorScaleNoDefault',
+  })(theme)
+  expect(result).toEqual({
+    color: {
```
Might be good to add a comment to state that this (although expected) won't result in valid CSS, in case someone stumbles upon this test

jxnblk commented on packages/css/test/index.ts:
```diff
@@ -206,13 +217,44 @@ test('works with functional arguments', () => {
+    color: 'color100',
+  })
+})
+
```
It'd be nice to have an additional test showing this working with a variant and returning an object

jxnblk reviewed (approved):
Left a few nits/comments, but nothing blocking -- this looks great, thanks!

folz committed:
Changes per PR review comments

- Update color scale to use actual colors
- Include a test for variant object default behavior
- Leave comment indicating object values can't become valid CSS

folz commented:
Thanks for the review! I made the changes you suggested. I'll open a new issue/PR about themeGet functions in themes.

folz commented:
Opened an issue to track `themeGet` reference support here: https://github.com/system-ui/theme-ui/issues/961

jxnblk commented:
Sweet! Thanks, I think this looks good to go, but want to let it sit for a bit so others have time to review or comment

hasparus commented:
This is a really good PR. Why didn't we merge it?
I'll handle the conflicts.

hasparus committed:
Remove is-object dependency

hasparus committed:
Loosen the check, default can be a responsive tuple or an object

hasparus committed:
chore: rename master to develop and previous to stable

dependabot-preview[bot] committed:
chore(deps-dev): bump lint-staged from 10.5.1 to 10.5.3

Bumps [lint-staged](https://github.com/okonet/lint-staged) from 10.5.1 to 10.5.3.
- [Release notes](https://github.com/okonet/lint-staged/releases)
- [Commits](https://github.com/okonet/lint-staged/compare/v10.5.1...v10.5.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump babel-preset-gatsby from 0.6.0 to 0.8.0

Bumps [babel-preset-gatsby](https://github.com/gatsbyjs/gatsby/tree/HEAD/packages/babel-preset-gatsby) from 0.6.0 to 0.8.0.
- [Release notes](https://github.com/gatsbyjs/gatsby/releases)
- [Changelog](https://github.com/gatsbyjs/gatsby/blob/master/packages/babel-preset-gatsby/CHANGELOG.md)
- [Commits](https://github.com/gatsbyjs/gatsby/commits/babel-preset-gatsby@0.8.0/packages/babel-preset-gatsby)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump execa from 4.1.0 to 5.0.0

Bumps [execa](https://github.com/sindresorhus/execa) from 4.1.0 to 5.0.0.
- [Release notes](https://github.com/sindresorhus/execa/releases)
- [Commits](https://github.com/sindresorhus/execa/compare/v4.1.0...v5.0.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

hasparus committed:
docs: link to develop and stable branches in readme

dependabot-preview[bot] committed:
chore(deps): bump @emotion/react from 11.1.1 to 11.1.2

Bumps [@emotion/react](https://github.com/emotion-js/emotion) from 11.1.1 to 11.1.2.
- [Release notes](https://github.com/emotion-js/emotion/releases)
- [Changelog](https://github.com/emotion-js/emotion/blob/master/CHANGELOG.md)
- [Commits](https://github.com/emotion-js/emotion/compare/@emotion/react@11.1.1...@emotion/react@11.1.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

hasparus committed:
Merge pull request #1341 from system-ui/dependabot/npm_and_yarn/emotion/react-11.1.2

chore(deps): bump @emotion/react from 11.1.1 to 11.1.2

dependabot-preview[bot] committed:
chore(deps): [security] bump ini from 1.3.5 to 1.3.7

Bumps [ini](https://github.com/isaacs/ini) from 1.3.5 to 1.3.7. **This update includes a security fix.**
- [Release notes](https://github.com/isaacs/ini/releases)
- [Commits](https://github.com/isaacs/ini/compare/v1.3.5...v1.3.7)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

hasparus committed:
Merge pull request #1343 from system-ui/dependabot/npm_and_yarn/lint-staged-10.5.3

chore(deps-dev): bump lint-staged from 10.5.1 to 10.5.3

hasparus committed:
Merge pull request #1356 from system-ui/dependabot/npm_and_yarn/ini-1.3.7

chore(deps): [security] bump ini from 1.3.5 to 1.3.7

hasparus committed:
Merge pull request #1344 from system-ui/dependabot/npm_and_yarn/babel-preset-gatsby-0.8.0

chore(deps-dev): bump babel-preset-gatsby from 0.6.0 to 0.8.0

dependabot-preview[bot] committed:
chore(deps-dev): bump babel-jest from 26.1.0 to 26.6.3

Bumps [babel-jest](https://github.com/facebook/jest/tree/HEAD/packages/babel-jest) from 26.1.0 to 26.6.3.
- [Release notes](https://github.com/facebook/jest/releases)
- [Changelog](https://github.com/facebook/jest/blob/master/CHANGELOG.md)
- [Commits](https://github.com/facebook/jest/commits/v26.6.3/packages/babel-jest)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

hasparus committed:
Merge pull request #1342 from system-ui/dependabot/npm_and_yarn/babel-jest-26.6.3

chore(deps-dev): bump babel-jest from 26.1.0 to 26.6.3

hasparus committed:
Merge pull request #1340 from system-ui/dependabot/npm_and_yarn/execa-5.0.0

chore(deps-dev): bump execa from 4.1.0 to 5.0.0

hasparus committed:
Merge branch 'master' into develop

hasparus committed:
Merge branch 'develop' into get-default

hasparus commented:
@folz, I'm gonna change this to `__default` to ensure people are using it knowingly. It turned out to be a bit surprising.
